x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportGetDataFrameTransformsStatsAction.java

```diff
@@ -212,6 +209,56 @@ private void collectStatsForTransformsWithoutTasks(Request request,
             }
         );
 
-        dataFrameTransformsConfigManager.getTransformStats(transformsWithoutTasks, searchStatsListener);
+        dataFrameTransformsConfigManager.getTransformStoredDoc(transformsWithoutTasks, searchStatsListener);
+    }
+
+    private void populateSingleStoppedTransformStat(DataFrameTransformStoredDoc transform,
+                                                    ActionListener<DataFrameTransformCheckpointingInfo> listener) {
+        transformsCheckpointService.getCheckpointStats(transform.getId(), transform.getTransformState().getCheckpoint(),
```

Comment:
I think we are going to find that this will hurt performance in the future. I know refactoring `getCheckpointStats` would a monstrous undertaking, and probably will have to be done in another PR.

Reply:
Yes, true, I should probably write a new method that gets all the required checkpoint docs for all the stopped transforms in one step.

Also getting the "operations behind" for many transforms that share the same source indices is inefficient because the indices stats action sends messages to the nodes holding the shards for each index that's involved and we'd do that multiple times.  So a more efficient approach would be to get the indices stats for the union of all indices referenced in the `source` of all transforms we're going to return stats for, then pick the relevant bits out of that for each transform.

The approach in this PR at the moment is actually what's suggested in #42978 but further optimisation would certainly be possible.  Since this PR is targeted at 7.4 we have time to do this in a followup.  Otherwise like you say the title of this PR will have to be changed to "refactor everything".